orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -207,39 +280,274 @@ Status TrainingRunner::Run(IDataLoader* training_data_loader, IDataLoader* test_
   return Status::OK();
 }
 
-Status TrainingRunner::TrainingLoop(IDataLoader& training_data_loader, IDataLoader* test_data_loader) {
-  const bool enable_checkpoint_saving =
-      params_.mpi_context.world_rank == 0 &&
-      checkpoint_registry_ && params_.checkpoint_period > 0;
+// Prepare feeds for a call to one session run.
+Status TrainingRunner::PrepareFeedNamesAndFeeds(IDataLoader& training_data_loader,
+                                                LearningRateScheduler& lr_scheduler,
+                                                const size_t batch_index,
+                                                std::vector<std::string>& feed_names,
+                                                std::vector<MLValue>& feeds) {
+  // Initialize outputs of this function.
+  feed_names = std::vector<std::string>();
+  feeds = std::vector<MLValue>();
+
+  auto allowed_feed_begin = pipeline_context_.feed_names.begin();
+  auto allowed_feed_end = pipeline_context_.feed_names.end();
+
+  // Pick up feeds from data loader
+  {
+    auto training_data = training_data_loader.CurrentDataSet();
+    std::vector<std::string> data_feed_names = training_data_loader.DataSetTensorNames();
+    std::vector<MLValue> data_feeds = training_data->GetKthBatch(params_.batch_size, batch_index, input_allocator_);
+    for (size_t i = 0; i < data_feed_names.size(); ++i) {
+      const auto name = data_feed_names[i];
+      if (params_.num_pipeline_stages == 1 || std::find(allowed_feed_begin, allowed_feed_end, name) != allowed_feed_end) {
+        feed_names.push_back(name);
+        feeds.push_back(data_feeds[i]);
+      }
+    }
+  }
 
-  VectorString feed_names = training_data_loader.DataSetTensorNames();
+  // Pick up feed from loss scaling.
   if (loss_scaler_) {
-    feed_names.push_back(loss_scaler_->GetLossScaleInputName());
+    const auto name = loss_scaler_->GetLossScaleInputName();
+    if (params_.num_pipeline_stages == 1 || std::find(allowed_feed_begin, allowed_feed_end, name) != allowed_feed_end) {
```

Comment:
>params_.num_pipeline_stages == 1 || [](start = 8, length = 35)

Is this needed? Could this feed show up in non-last stage of pipeline?

Reply:
In most cases, `loss_scaler` only appears in the last stage, so we omit out it if middle stages don't  allow it as a feed.

Reply:
Yes that's why I think the check for pipeline_parallel_size == 1 is not needed here, since the check for loss_scaler is independent of whether pipeline is used or not.

---
In reply to: [419612450](https://github.com/microsoft/onnxruntime/pull/3770#discussion_r419612450) [](ancestors = 419612450)

Reply:
I am guessing pipeline_context_.feed_names only contain graph input names, not including loss_scaler name and lr name? 

---
In reply to: [420978149](https://github.com/microsoft/onnxruntime/pull/3770#discussion_r420978149) [](ancestors = 420978149,419612450)